### DISCUSSION TITLE
Impact of dependency injection on performances

### DISCUSSION THREAD

**Discussion by VincentLanglet** [symfony/symfony#45187] [Q&A] [answered]
Hi,

I recently joined a team with a big legacy project. In this one, we often have a 30 second timeout when generating the dev cache, and there was some circular references which I think was solved with trick like `@required`, closure, calling service of service, ...

They have the following "best" practices:
- Prefer one big service over multiple small one: it avoids service injection
- Prefer static method: it avoids service injection
- All service injection must be done with `@required` rather than `__construct` methods.

The best argument given is "Service injection has bad impact on performances".

I kinda disagree on this, but I don't have a lot of knowledge about it. Could I get some explanations/arguments ?
Thanks !

**Comment by stof** [comment, marked as answer]
What has a performance impact **in debug mode** is the usage of autowiring and maybe also of autoconfiguration (compared to defining services explicitly in config files), because it forces to recheck the cached container each time a PHP file changes (even though that change did not impact the dependencies of the service) while an explicit config invalidates the cached container only when you edit that XML or Yaml config file (but if the container actually needs a rebuild anyway due to the change, the impact is not high IIRC).

Circular references might make things slower during the analysis (to detect whether they are acceptable), in which case the best solution is to actually avoid them in the object graph instead of hoping that those circular dependencies will end up being an accepted case for the instantiation. Most services don't need to have circular dependencies in my experience, especially if we avoid mixing responsibilities in the same service.
`@required` setters vs the constructor does not change anything to me regarding the performance of the DI component (and using setters implies that your classes are mutable). The case where using setter might help with is making circular object graphs acceptable instead of being an error by making them instantiable. But the preferred solution is still to try to break the cycle instead, or using that when needed for that rather than all the time.

```php
// circular dependency that can be instantiated with setters
$a = new A();
$b = new B();
$b->setA($a);
$a->setB($b);

// circular dependency that is impossible to instantiate
$c = new C($d); // $d does not exist yet...
$d = new D($c);
```

From my experience, the biggest performance issue impacting the DI container building (and especially when using autowiring and autoconfiguration that involves all your PHP files as resources, which means a lot more files) is the IO performance of your disk. Putting the source code in a virtualbox shared folder gives an insane time for the cache generation, while putting the source code on an SSD drive makes that very fast, and quite far from the 30s timeout.

> Prefer static method: it avoids service injection

This rules means you are not using DI anymore, and so you loose all its benefits (and it won't be an option when interacting with any code from Symfony or open-source packages that are meant to be used in a DI-based environment)

**Reply by VincentLanglet** [reply]
> What has a performance impact **in debug mode** is the usage of autowiring and maybe also of autoconfiguration (compared to defining services explicitly in config files), because it forces to recheck the cached container each time a PHP file changes (even though that change did not impact the dependencies of the service) while an explicit config invalidates the cached container only when you edit that XML or Yaml config file (but if the container actually needs a rebuild anyway due to the change, the impact is not high IIRC).

So can you confirm, that
- Moving code into a service and injecting it
- Splitting a service into multiple smaller ones, and injecting them
has no impact on performances ? (If we're excluding debug mode)

I consider this as best practices and would like to promote them.
 
> Circular references ...

I fully agree that circular references should be solved instead

> > Prefer static method: it avoids service injection
> 
> This rules means you are not using DI anymore, and so you loose all its benefits (and it won't be an option when interacting with any code from Symfony or open-source packages that are meant to be used in a DI-based environment)

They inject as few services as possible so there is still some DI.
But this rule means that unit tests are a mess since you can't mock a static method.

**Reply by stof** [reply]
@VincentLanglet as long as constructors are not doing heavy logic, this is fine.

If a class does heavy logic in its constructor (looking at you PDO), then it is a bad fit for being defined as a service (or least for being injected into other services, which is the primary usage of services), in which case marking the service as lazy *might* help.
